spec/helpers/faker/puppet_module_remote_repo.rb

```diff
@@ -0,0 +1,140 @@
+  module Faker
+    class PuppetModuleRemoteRepo
+      class CommandExecutionError < StandardError; end
+
+      attr_reader :name
+      attr_reader :namespace
+
+      def initialize(name, namespace)
+        @name = name
+        @namespace = namespace
+      end
+
+      def populate
+        FileUtils.chdir(Faker.working_directory) do
+          run %W{git init --bare #{bare_repo_dir}}
+          run %W{git clone #{bare_repo_dir} #{tmp_repo_dir}}
+
+          module_short_name = name.split('-').last
+
+          FileUtils.chdir(tmp_repo_dir) do
+            metadata = {
+              name: "modulesync-#{module_short_name}",
+              version: '0.4.2',
+              author: 'ModuleSync team',
+            }
+
+            File.write 'metadata.json', metadata.to_json
+            run %w{git add metadata.json}
+            run %W{git commit --message} << 'Initial commit'
+            run %w{git push}
+          end
+        end
+      end
+
+      def read_only=(value)
+        mode = value ? '0444' : '0644'
+        FileUtils.chdir(bare_repo_dir) do
+          run %W{git config core.sharedRepository #{mode}}
+        end
+      end
+
+      def default_branch
+        FileUtils.chdir(bare_repo_dir) do
+          stdout = run %w{git symbolic-ref --short HEAD}
+          return stdout.chomp
+        end
+      end
+
+      def default_branch=(value)
+        FileUtils.chdir(bare_repo_dir) do
+          run %W{git branch -M #{default_branch} #{value}}
+          run %W{git symbolic-ref HEAD refs/heads/#{value}}
+        end
+      end
+
+      def read_file(filename, branch = nil)
+        branch ||= default_branch
+        FileUtils.chdir(bare_repo_dir) do
+          return run %W{git show #{branch}:#{filename}}
+        rescue CommandExecutionError
+          return nil
+        end
+      end
+
+      def add_file(filename, content, branch = nil)
+        branch ||= default_branch
+        FileUtils.chdir(tmp_repo_dir) do
+          run %W{git checkout #{branch}}
+          File.write filename, content
+          run %W{git add #{filename}}
+          run %W{git commit --message} << "Add file: '#{filename}'"
+          run %w{git push}
+        end
+      end
+
+      def commit_count_between(commit1, commit2)
+        FileUtils.chdir(bare_repo_dir) do
+          stdout = run %W{git rev-list --count #{commit1}..#{commit2}}
+          return Integer(stdout)
+        end
+      end
+
+      def commit_count_by(author, commit = nil)
+        FileUtils.chdir(bare_repo_dir) do
+          commit ||= '--all'
+          stdout = run %W{git rev-list #{commit} --author #{author} --count}
+          return Integer(stdout)
+        end
+      end
+
+      def remote_url
+        "file://#{bare_repo_dir}"
+      end
+
+      def self.git_base
+        "file://#{Faker.working_directory}/bare/"
+      end
+
+      private
+      def tmp_repo_dir
+        "#{Faker.working_directory}/tmp/#{namespace}/#{name}"
```

Comment:
Not sure how much we care about other platforms, but might be nice to use `File.join`.

Reply:
Answer is in question: do we really care about other platforms? ;-)

Reply:
I've learned to use it also a way to signal it's a path. An alternative is `Pathname`. I think you can make `Faker.working_directory` a `Pathname` instance and use `Faker.working_directory / 'tmp' / namespace / name` instead.

Reply:
I trend to think that File#join is more readable than Pathname

Reply:
OK